### ISSUE TITLE
generate_heatmap crashes notebooks

### ISSUE THREAD

**Issue by HinePo** [ultralytics/ultralytics#6892]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

Hi, Ultralytics team,

I am having some troubles using the new heatmap feature on notebooks (Google Colab and Kaggle notebooks).

From what I've searched, this is due to the  cv2.imshow() function, not supported in notebooks.

The script that fails is [ultralytics/solutions/heatmap.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/solutions/heatmap.py), on line 171: `cv2.imshow('Ultralytics Heatmap', im0_with_heatmap)`

I wonder if this can be fixed so we can use the new heatmap feature in notebooks.

Many thanks to all Ultralytics team for the awesome work and lib, as always!

![image](https://github.com/ultralytics/ultralytics/assets/66163270/2a26c975-0d78-4639-8573-769a6421a75f)

### Environment

Ultralytics YOLOv8.0.225 🚀 Python-3.10.12 torch-2.1.0+cu118 CUDA:0 (Tesla T4, 15102MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 27.1/166.8 GB disk)

OS                  Linux-5.15.120+-x86_64-with-glibc2.35
Environment         Colab
Python              3.10.12
Install             pip
RAM                 12.68 GB
CPU                 Intel Xeon 2.30GHz
CUDA                11.8

matplotlib          ✅ 3.7.1>=3.3.0
numpy               ✅ 1.23.5>=1.22.2
opencv-python       ✅ 4.8.0.76>=4.6.0
pillow              ✅ 9.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.11.4>=1.4.1
torch               ✅ 2.1.0+cu118>=1.8.0
torchvision         ✅ 0.16.0+cu118>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
pandas              ✅ 1.5.3>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1

### Minimal Reproducible Example

```

### https://docs.ultralytics.com/guides/heatmaps/#__tabbed_1_4

from ultralytics import YOLO
from ultralytics.solutions import heatmap
import cv2

model = YOLO("yolov8s.pt")   # YOLOv8 custom/pretrained model

cap = cv2.VideoCapture("path/to/video/file.mp4")  # Video file Path, webcam 0
assert cap.isOpened(), "Error reading video file"

# Region for object counting
count_reg_pts = [(20, 400), (1080, 404), (1080, 360), (20, 360)]

# Heatmap Init
heatmap_obj = heatmap.Heatmap()
heatmap_obj.set_args(colormap=cv2.COLORMAP_JET,
                     imw=cap.get(4),  # should same as im0 width
                     imh=cap.get(3),  # should same as im0 height
                     view_img=True,
                     count_reg_pts=count_reg_pts)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        exit(0)
    results = model.track(im0, persist=True)
    im0 = heatmap_obj.generate_heatmap(im0, tracks=results)

```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by RizwanMunawar** [comment]
@HinePo Thanks for providing the information. Yes, I can reproduce your issue. It will be fixed soon! 

Thanks
Ultralytics Team!

**Comment by HinePo** [comment]
Many thanks, @RizwanMunawar and all Ultralytics team! I think it is much better now, the generate_heatmap is not crashing on that step now.

But maybe there is still room for improvements. I've noticed that:

- at the end of the generate_heatmap processing, it still crashes the notebooks (Colab and Kaggle), after all processing is done. I think it is a problem with the last frame in the loop, that returns None Type and crashes opencv. Previously, it crashed at the beginning of the method, this is now fixed. But now it crashes at the end of the execution
- So I had to add a try/exception, but it only resolves this issue when I use bytetrack. When using botsort this fix does not work
- Saving results (detected video) in runs/detect/predict as usual is not working well for heatmaps. It does not save the video (maybe because it crashes right before the end of the generate_heatmap?). But it saves the txt file correctly under the name of `image0.txt` as expected
- Some Predict/Track arguments like `show_conf`, `show_boxes`, `save` do not seem to work with the heatmap method, I don't know if this is intended. But following the documentation I could save the detected video using video writer, and found the results amazing!
- By the way, is there a tolerance threshold (in frames or seconds) to tune the tracker/InCount/OutCount? This feature is awesome and saves so much code and work!!

Anyway, **many thanks again** to all Ultralytics team for the lightining fast answers and work!!
